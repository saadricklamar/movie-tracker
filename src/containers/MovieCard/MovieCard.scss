.movie-card {
  position: relative;
  overflow: hidden;
  perspective: 1000px;
  padding: 1.5rem;
  cursor: pointer;
  align-items: flex-start;
  box-shadow: 0px 7px 10px rgba(black, 0.5);
}

.fa-star {
  color: white;
  font-size: 25px;
  margin: 10px;
}

.fa-star:hover {
  color: gold;
}

.favorite {
  color: gold;
}

.movie-card:hover .card,
.movie-card.hover .card {
  transform: rotateY(180deg);
}

.movie-card,
.front,
.back {
  width: 300px;
  height: 400px;
}

.card {
  transition: 0.6s;
  transform-style: preserve-3d;
  position: relative;
}

.front,
.back {
  backface-visibility: hidden;
  position: absolute;
  top: 0;
  left: 0;
}

.front {
  z-index: 2;
  background-position: center;
  background-repeat: no-repeat;
  background-size: cover;
  transform: rotateY(0deg);
  border: 1px solid white;
  border-radius: 15px;
}

.back {
  transform: rotateY(180deg);
  border: 1px solid white;
  border-radius: 15px;
  text-align: left;
}

.back p {
  color: white;
}

.overview {
  margin-top: 10px;
  padding: 5px;
  font-size: 13px;
  font-weight: bold;
  transition: all 0.2s ease-in-out;
}

.title,
.rating,
.release {
  font-family: "Poiret One";
  font-size: 20px;
  padding: 5px 10px 0 10px;
}

.title {
  font-size: 24px;
  color: white;
}
