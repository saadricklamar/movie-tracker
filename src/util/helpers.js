export const cleanMovieData = movies => {
  return movies.results.map(movie => {
    return {
      title: movie.title || movie.name,
      rating: movie.vote_average,
      id: movie.id,
      releaseDate: movie.release_date,
      synopsis: movie.overview,
      posterImage:
        "https://image.tmdb.org/t/p/w185_and_h278_bestv2" + movie.poster_path,
      favorite: false
    };
  });
};
