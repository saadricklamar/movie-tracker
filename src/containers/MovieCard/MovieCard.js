import React, { Component } from "react";
import { PropTypes } from "prop-types";
import { connect } from "react-redux";
import * as actions from "../../util/fetchData";
import { toggleFavorite } from "../../actions";

export class MovieCard extends Component {
  constructor() {
    super();
    this.state = {
      favorite: false
    };
  }

  componentDidMount = async () => {
    const currentFavorites = await actions.getFavorites(this.props.user_id);
    const locatedFavorites = currentFavorites.data.find(favoriteMovie => {
      return favoriteMovie.movie_id === this.props.movie.id;
    });
    if (locatedFavorites) {
      this.setState({ favorite: true });
    } else {
      this.setState({ favorite: false });
    }
  };

  toggleMovieFavorite = async id => {
    this.props.toggleFavorite(id);
    let { movie, user_id } = this.props;
    let movieID = movie.id;
    if (this.state.favorite) {
      movieID = movie.movie_id;
    }
    if (this.state.favorite) {
      await actions.deleteFavorite(user_id, movie.id);
    } else {
      await actions.addFavorite(movie.id, user_id, movie);
    }
    this.setState({ favorite: !this.state.favorite });
  };

  render() {
    return (
      <div className="movie-card">
        <div className="card">
          <div
            className="front"
            style={{ backgroundImage: `url(${this.props.movie.poster_path})` }}
          />
          <div className="back">
            <i
              onClick={() => this.toggleMovieFavorite(this.props.movie.id)}
              className={`fas fa-star ${this.state.favorite && "favorite"}`}
            />
            <p className="title">{this.props.movie.title}</p>
            <p className="rating">
              Rating: {this.props.movie.vote_average} / 10
            </p>
            <p className="release"> Release: {this.props.movie.release_date}</p>
            <p className="overview scroller">
              Summary: {this.props.movie.overview}
            </p>
          </div>
        </div>
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  user_id: state.users.id
});

export const mapDispatchToProps = dispatch => ({
  toggleFavorite: id => dispatch(toggleFavorite(id))
});

MovieCard.propTypes = {
  movie: PropTypes.object.isRequired,
  user_id: PropTypes.number.isRequired
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MovieCard);
