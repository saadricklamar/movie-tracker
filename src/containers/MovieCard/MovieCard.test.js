import React from 'react';
import ReactDOM from 'react-dom';
import {MovieCard} from './MovieCard';
import { shallow, mount, render } from 'enzyme';
import {mapStateToProps} from  './MovieCard'
import * as actions from '../../util/fetchData';
import { Provider } from 'react-redux';

jest.mock('../../util/fetchData')


describe("MovieCard", () => {
  let wrapper;
  let mockMovie;
  let enzyme = require("enzyme");

  beforeEach(() => {
    mockMovie = {
      title: "Avengers",
      vote_average: 10,
      id: 1,
      release_date: 2019,
      overview: "a darn good movie",
      poster_path: undefined
    };
    wrapper = enzyme.shallow(<MovieCard movie={mockMovie} user_id={1} />);
  });

  describe("MovieCard", () => {
    it("should match the snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });

    describe("ComponentDidMount", () => {
      it("should call getFavorites on componentDidMount", async () => {
        actions.getFavorites.mockImplementation(() => ({ data: [mockMovie] }));
        wrapper = shallow(<MovieCard movie={mockMovie} user_id={1} />, {
          disableLifecycleMethods: true
        });
        await wrapper.instance().componentDidMount();
        expect(actions.getFavorites).toHaveBeenCalled();
      });

      it("set favorite to true if its a current favorite", async () => {
        actions.getFavorites.mockImplementation(() => ({
          data: [{ movie_id: 1 }]
        }));
        wrapper = shallow(<MovieCard movie={mockMovie} user_id={1} />, {
          disableLifecycleMethods: true
        });
        await wrapper.instance().componentDidMount();
        expect(wrapper.state().favorite).toBe(true);
      });

      it("set favorite to false if its a current favorite", async () => {
        actions.getFavorites.mockImplementation(() => ({
          data: [{ movie_id: 2 }]
        }));
        wrapper = shallow(<MovieCard movie={mockMovie} user_id={1} />, {
          disableLifecycleMethods: true
        });
        await wrapper.instance().componentDidMount();
        expect(wrapper.state().favorite).toBe(false);
      });
    });
  });
});
