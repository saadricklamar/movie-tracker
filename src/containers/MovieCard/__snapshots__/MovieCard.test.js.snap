// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard MovieCard should match the snapshot 1`] = `
<div
  className="movie-card"
>
  <div
    className="card"
  >
    <div
      className="front"
      style={
        Object {
          "backgroundImage": "url(undefined)",
        }
      }
    />
    <div
      className="back"
    >
      <i
        className="fas fa-star false"
        onClick={[Function]}
      />
      <p
        className="title"
      >
        Avengers
      </p>
      <p
        className="rating"
      >
        Rating: 
        10
         / 10
      </p>
      <p
        className="release"
      >
         Release: 
        2019
      </p>
      <p
        className="overview scroller"
      >
        Summary: 
        a darn good movie
      </p>
    </div>
  </div>
</div>
`;
